### PR TITLE
BufferList slice/negative-index support

### LINKED ISSUE #471: Bug: Hierarchical up-graph visualization uses wrong graph object (pyg_down instead of pyg_up)
In `create_graph.py`, when `create_plot=True` and `hierarchical=True`, the up-graph visualization incorrectly passes `pyg_down` to `plot_graph` instead of `pyg_up`:

https://github.com/mllam/neural-lam/blob/8141679/neural_lam/create_graph.py#L345-L347

```python
# Current (bug): plots the down graph twice
plot_graph(pyg_down, title=f"Up graph, {to_level} -> {from_level}")

# Should be:
plot_graph(pyg_up, title=f"Up graph, {to_level} -> {from_level}")
```

# Impact

Since plot_graph colors nodes by in-degree (pyg.utils.degree(edge_index[1])), and pyg_up has reversed edge indices compared to pyg_down, the "Up graph" visualization shows the wrong in-degree pattern, it displays the down-graph's receiver-degree distribution instead of the up-graph's.

Anyone debugging hierarchical graph construction with --plot gets a misleading visualization where up and down graphs appear identical.

#Additional improvement

While investigating, I also noticed that BufferList.__getitem__ only supports positive integer indexing, forcing list() conversions in BaseHiGraphModel.process_step to enable slicing. I've added slice and negative-index support following the Python sequence protocol, consistent with nn.ModuleList behavior.

### PR DESCRIPTION
**Pull request by sudhansu-24**
## Describe your changes

Add Python sequence protocol support to `BufferList` and clean up downstream usage.

**BufferList ergonomics / minor performance cleanup**
- Extend BufferList.__getitem__ to support Python slicing (buffer_list[1:], buffer_list[-2:], step slices) and negative indices.

**Downstream cleanup**
- Remove redundant list(...) wrapping in BaseHiGraphModel.process_step by relying on the new slice behavior.

Tests
- Add unit tests for BufferList integer indexing, negative indexing, slicing semantics, and iteration.

**Motivation / context**
`BufferList` is used throughout the hierarchical graph models to store per-level tensors. Supporting slicing and negative indices makes hierarchical processing code more Pythonic and eliminates the need for explicit `list()` conversions, matching standard Python container expectations.

*The `create_graph.py` visualization fix from the original PR has been dropped per @sadamov's guidance, that code path will be superseded by the weather-model-graphs migration (#83).*

**Dependencies**
No new dependencies.

## Issue Link

N/A (split from #471 — the visualization portion will be resolved by #83)

## Type of change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [ ] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [x] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by sadamov** [comment]
@sudhansu-24 thanks, similar to other PRs about the `create_graph` script I am here to warn you that we plan to soon move to weather-model-graph #83 for graph creation. That repo has beautiful (and hopefully accurate ;)) plotting in place. unless you have a strong preference to get this fixed right now, I propose to close this PR and wait for #83, let me know what you think.

**Comment by sudhansu-24** [comment]
Hi @sadamov , thanks for the heads up! 
I wasn't aware of the plan to move to weather-model-graphs (#83) that makes total sense.

I'm happy to drop the create_graph.py visualization fix since wmg already handles that properly. However, the BufferList slice/negative-index support (in utils.py) & coresponding cleanup in base_hi_graph_model.py are unrelated to create_graph would it make sense to split those into a separate PR, or would you prefer to close this entirely??

**Comment by sadamov** [comment]
@sudhansu-24 that makes sense, let's focus this one #472 on the bufferlist then

**Comment by sudhansu-24** [comment]
@sadamov  Updated #472  to focus only on BufferList slice/negative-index support and the related unit tests. removed the create_graph.py visualization changes now, marked the PR as ready for review

File: neural_lam/utils.py [hunk 1]
```diff
@@ -36,6 +36,10 @@ def __init__(self, buffer_tensors, persistent=True):
             self.register_buffer(f"b{buffer_i}", tensor, persistent=persistent)
 
     def __getitem__(self, key):
+        if isinstance(key, slice):
+            return [self[i] for i in range(*key.indices(len(self)))]
+        if key < 0:
+            key += len(self)
         return getattr(self, f"b{key}")
```
Comment (sadamov) [inline neural_lam/utils.py:63]: Bug: out-of-bounds negative index raises `AttributeError` instead of `IndexError`.

`key += len(self)` only runs once, so `buf[-100]` on a 5-element list yields `key = -95` and `getattr(self, "b-95")` raises `AttributeError`. Python sequences must raise `IndexError`.

```suggestion
        if key < 0:
            key += len(self)
        if not (0 <= key < len(self)):
            raise IndexError(
                f"index {key} out of range for BufferList of length {len(self)}"
            )
        return getattr(self, f"b{key}")
```

**Review by sadamov — COMMENTED** [review]
No CHANGELOG entry added (author checklist item unchecked).

File: tests/test_graph_creation.py [hunk 2]
```diff
@@ -117,3 +118,60 @@ def test_graph_creation(datastore_name, graph_name):
...
+
+    def test_integer_index(self, buffer_list):
+        """Positive integer indexing returns the correct buffer."""
+        assert torch.equal(buffer_list[0], torch.tensor([0.0]))
+        assert torch.equal(buffer_list[4], torch.tensor([4.0]))
+
+    def test_negative_index(self, buffer_list):
+        """Negative indexing follows Python sequence convention."""
+        assert torch.equal(buffer_list[-1], torch.tensor([4.0]))
+        assert torch.equal(buffer_list[-3], torch.tensor([2.0]))
```
Comment (sadamov) [inline tests/test_graph_creation.py]: Missing test for the out-of-bounds bug above.

```suggestion
    def test_negative_index(self, buffer_list):
        """Negative indexing follows Python sequence convention."""
        assert torch.equal(buffer_list[-1], torch.tensor([4.0]))
        assert torch.equal(buffer_list[-3], torch.tensor([2.0]))

    def test_out_of_bounds_raises(self, buffer_list):
        """Out-of-bounds access raises IndexError, not AttributeError."""
        with pytest.raises(IndexError):
            buffer_list[5]
        with pytest.raises(IndexError):
            buffer_list[-6]
```

File: tests/test_graph_creation.py [hunk 3]
```diff
@@ -117,3 +118,60 @@ def test_graph_creation(datastore_name, graph_name):
                         assert r.shape[0] == 2  # adjacency matrix uses two rows
                     elif file_id.endswith("_features"):
                         assert r.shape[1] == d_features
+
+
+class TestBufferList:
```
Comment (sadamov) [inline tests/test_graph_creation.py]: `TestBufferList` is the only test class in the repo — all other tests use plain `def test_*` functions. Please flatten to match the convention.

**Comment by sudhansu-24** [comment]
Thanks for the review. i have added the IndexError guard for out-of-bounds indices, expanded tests, flattened the BufferList tests and added a CHANGELOG entry under unreleased.

**Review by sadamov — APPROVED** [review]
LGTM

File: neural_lam/utils.py [hunk 4]
```diff
@@ -36,6 +36,14 @@ def __init__(self, buffer_tensors, persistent=True):
             self.register_buffer(f"b{buffer_i}", tensor, persistent=persistent)
 
     def __getitem__(self, key):
+        if isinstance(key, slice):
+            return [self[i] for i in range(*key.indices(len(self)))]
+        if key < 0:
+            key += len(self)
```
Comment (joeloskarsson) [inline neural_lam/utils.py:58]: Could be nice to add a couple comments just explaining this, that we are 1. unpacking the indices in the slice and calling same method recursively and 2. make sure negative indices work and index from the back.

**Review by joeloskarsson — APPROVED** [review]
Could add some clarifying comment, but otherwise good. Thank for this addition! Looks good to me.

File: neural_lam/utils.py (hunk 4 above)
Comment (sudhansu-24) [reply neural_lam/utils.py:58]: sure, now i have added comments for slice handling and negative indexing.

**Comment by sadamov** [comment]
okay great, then we wait for https://github.com/mllam/neural-lam/pull/208 to be merged before we merge this one! :raised_hands:

**Comment by sadamov** [comment]
#208 was just merged, could you resolve conflicts here @sudhansu-24 so that we can merge the PR?

**Comment by sudhansu-24** [comment]
> #208 was just merged, could you resolve conflicts here @sudhansu-24 so that we can merge the PR?

done, rebased onto latest `main` and resolved merge conflicts in `utils.py` and  `CHANGELOG.md`

**Comment by joeloskarsson** [comment]
Still lgtm, just need to make tests and linting pass

**Review by sadamov — APPROVED** [review]
thanks again @sudhansu-24 this looks good now, merging!
